### DISCUSSION TITLE
SENS_SPI_FRAMING_ERROR & SENS_SPI_PARITY_ERROR When write data into SETTINGS2(0x0019) register

### DISCUSSION THREAD

**Discussion by wish029517** [jonas-merkle/AS5047P#46] [Q&A]
I want to set AS5047D  SETTINGS2(0x0019) register like the following command:
UVWPP : 011 (4 of pole pairs)
HYS: 00
ABIRES: 000

When I write 0x0003 into the register,I got "SENS_SPI_FRAMING_ERROR" 、 "SENS_SPI_PARITY_ERROR" and "CONT_SPI_PARITY_ERROR",
but according to even parity,I think it is not wrong.

This is my code:
  auto settingsSETTINGS1 = new AS5047P_Types::SETTINGS1_t();
  auto settingsSETTINGS2 = new AS5047P_Types::SETTINGS2_t();
  
  // initialize a new object which represents error information
  auto errorInfo = AS5047P_Types::ERROR_t(); 

  // set the raw register value
  settingsSETTINGS1->data.raw = 0x0088;
  settingsSETTINGS2->data.raw = 0x0003;

  // write the settings 1、2 register to the sensor
  bool SETTINGS1_result = as5047p.write_SETTINGS1(settingsSETTINGS1, &errorInfo, true, true);
  bool SETTINGS2_result = as5047p.write_SETTINGS2(settingsSETTINGS2, &errorInfo, true, true);

Thank you very much for help!

**Comment by jonas-merkle** [comment]
Hi @wish029517,
could you please provide me the version of the library and the type of hardware you're using?
Since you also get an SPI framing error, the following error sources come to mind:
- The Arduino starts sending data so fast that the sensor is not ready for it. Adding a `delay(1000);` before writing the registers might help.
- The Arduino was reset during the communication with the sensor. This may lead to an undefined state of the sensor, so that it has to be reset as well.

maybe this already helps.

**Comment by wish029517** [comment]
The version of library is 2.2.0,and the type of hardware is Arduino MEGA 2560.
The problem is still unsolved.There is one thing comfused me, the SPI framing error only appears when I write SETTINGS2.....

**Reply by jonas-merkle** [reply]
Just so I understand the problem correctly.... Writing to register SETTINGS1 works fine and when writing to register SETTINGS2 an error occurs? 

At the moment I have no idea what could cause such an error. But you can try to write into the SETTINGS2 register with deactivated error detection and then check manually if the values are written correctly... see code snippet:

```cpp
// your setup code ...

// init new SETTINGS2 object representing the register
auto settings = new AS5047P_Types::SETTINGS2_t();
settings->data.values.UVWPP = 3;
settings->data.values.HYS = 0;
settings->data.values.ABIRES = 0;

// write the SETTINGS2 object to the sensor without any error detection
as5047p.write_SETTINGS2(settings);

// wait a second
delay(1000);

// read the content of the SETTINGS2 register
auto res = as5047p.read_SETTINGS2();

// compare the settings object an the res object ...
```
